lms/services/canvas_api/client.py

```diff
@@ -226,16 +227,31 @@ class _ListFilesSchema(RequestsResponseSchema):
         id = fields.Integer(required=True)
         updated_at = fields.String(required=True)
 
-    def public_url(self, file_id):
+    def public_url(self, course_id, file_id, is_instructor):
         """
         Get a new temporary public download URL for the file with the given ID.
 
+        :param course_id: the ID of the Canvas course
+        :rtype: str
+
         :param file_id: the ID of the Canvas file
         :rtype: str
+
+        :param is_instructor: Whether the current user is an instructor
+        :type is_instructor: bool
+
+        :raise CanvasFileNotFoundInCourse: If there's no file with ID file_id
+            in the course with ID course_id
         """
         # For documentation of this request see:
         # https://canvas.instructure.com/doc/api/files.html#method.files.public_url
 
+        if is_instructor:
+            if not any(
+                str(file_["id"]) == file_id for file_ in self.list_files(course_id)
+            ):
+                raise CanvasFileNotFoundInCourse(file_id)
+
```

Comment:
I've moved this to a separate `Client.check_file_in_course()` method, which simplifies the `Client` tests nicely and keeps `is_instructor` out of the `Client` service.

This is also inline with Rob's suggestion of moving it to an entirely separate API / view. Once we do that the `course_id` will no longer need to be included in the `via_url` API request at all, so it'll be even cleaner. And it'll be faster because we'll have more parallelism, and the chances of request timeouts will be lower as we'll be making fewer external requests per request to us. But that's for a follow-up PR.

Reply:
One observation: this meant that the test that if the user is an instructor it checks that the file is in the course got moved into the view's tests. Since the view uses a mock `canvas_api_client`, this test was also much easier to write:

```python
@pytest.mark.usefixtures("user_is_instructor")
def test_it_checks_whether_the_file_is_in_the_course(self, canvas_api_client, pyramid_request):
    FilesAPIViews(pyramid_request).via_url()

    canvas_api_client.check_file_in_course.assert_called_once_with("test_file_id", "test_course_id")
```

Checking what it does when the file _isn't_ in the course was easy too, as to create that situation I only need to do `     canvas_api_client.check_file_in_course.side_effect = CanvasFileNotFoundInCourse()`.

Because the Canvas API client's tests don't use mocks, the test was much harder to write when the logic was in the Canvas API client. The test has to know what Canvas API responses look like even though that doesn't have anything to do with `check_file_in_course()`'s code (this knowledge about what Canvas API responses looks like ends up duplicated in many places in the tests). The assertion is particularly egregious as we have to make an assertion about a network request that's made in `_basic.py` (which is called by `_authenticated.py`, which is called by `client.py`). All the method under test actually does is call `client.send()` but we can't assert about that because it's a real `AuthenticatedClient` not a mock:

```python
http_session.send.return_value = factories.requests.Response(
    json_data=[
        {
            "display_name": "display_name_1",
            "id": 1,
            "updated_at": "updated_at_1",
        },
        {
            "display_name": "display_name_2",
            "id": 2,
            "updated_at": "updated_at_2",
        },
    ],
    status_code=200,
)

response = canvas_api_client.public_url(course_id="test_course_id", file_id="1", is_instructor=True)

http_session.send.assert_called_once_with(
    Any.request(
        url=Any.url.with_path("api/v1/courses/test_course_id/files"),
    ),
    timeout=Any()
)
```

 It's much harder to tell what's wrong when this test fails as well, since the range of code that could contain the bug or that the exception could arise from is much wider. For the same reason it's much harder to tell whether the test is correct and is in fact testing what it should be, not giving false positive test passes. Much harder to tell whether everything has been tested, or whether there are cases missing. The chances of the tests passing even though there is a bug in the code, are much higher. And the tests are much harder to understand and maintain.